underfs/hdfs/src/main/java/tachyon/underfs/hdfs/HdfsUnderFileSystemFactory.java

```diff
@@ -15,19 +15,45 @@
 
 package tachyon.underfs.hdfs;
 
+import java.util.Map;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
 import com.google.common.base.Preconditions;
+import com.google.common.collect.Maps;
 
 import tachyon.conf.TachyonConf;
 import tachyon.underfs.UnderFileSystem;
 import tachyon.underfs.UnderFileSystemFactory;
 
 public class HdfsUnderFileSystemFactory implements UnderFileSystemFactory {
+  /**
+   * Cache mapping {@link Path}s to existing {@link UnderFileSystem} instances. The Paths should be
+   * normalized to root paths because only their schemes and authorities are needed to identify
+   * which {@link FileSystem} they belong to.
+   */
+  private Map<Path, HdfsUnderFileSystem> mHdfsUfsCache = Maps.newHashMap();
 
   @Override
   public UnderFileSystem create(String path, TachyonConf tachyonConf, Object conf) {
     Preconditions.checkArgument(path != null, "path may not be null");
 
-    return new HdfsUnderFileSystem(path, tachyonConf, conf);
+    // Normalize the path to just its root. This is all that's needed to identify which FileSystem
+    // the Path belongs to.
+    Path rootPath = getRoot(new Path(path));
+    if (!mHdfsUfsCache.containsKey(rootPath)) {
+      mHdfsUfsCache.put(rootPath, new HdfsUnderFileSystem(path, tachyonConf, conf));
+    }
+    return mHdfsUfsCache.get(rootPath);
+  }
```

Comment:
It would need to accept a function which would lazily create the `HdfsUnderFileSystem`, or we wouldn't have any performance gain. The options are either synchronizing on the cache, or making the cache a `ConcurrentHashMap`. I like synchronizing on `mHdfsUfsCache` better because it's conceptually simpler (no need to worry about threads interleaving between the 3 `Map` methods), though it is slightly less efficient to hold the lock while creating `HdfsUnderFileSystem`.